### ISSUE TITLE
Operators that return dynamic-shape outputs that require_grad choke in AOTAutograd / unbacked SymInts are reallocated

### ISSUE THREAD

**Issue by williamwen42** [pytorch/pytorch#111950]
Custom autograd functions that use `impl_abstract` that give outputs with dynamic shapes are not supported in PT2.

Repro:
```python
import torch
from torch.library import Library, impl
from torch.testing._internal.optests import aot_autograd_check

m = Library("my_custom_library", "DEF")
m.define("custom_op(Tensor input, Tensor lengths) -> Tensor")

class CustomOp(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, lengths):
        ctx.save_for_backward(input, lengths)
        with torch._C._AutoDispatchBelowAutograd():
            return torch.ops.my_custom_library.custom_op(input, lengths)

    @staticmethod
    def backward(ctx, grad):
        input, lengths = ctx.saved_tensors
        idx = lengths.sum()
        grad_out = torch.zeros_like(input)
        grad_out[:idx] = grad
        return grad_out, None

@impl(m, "custom_op", "CPU")
def custom_op_impl(input, lengths):
    idx = lengths.sum()
    return input[:idx].clone()

@impl(m, "custom_op", "Autograd")
def custom_op_autograd(input, lengths):
    return CustomOp.apply(input, lengths)

@torch.library.impl_abstract("my_custom_library::custom_op")
def custom_op_meta(input, lengths):
    return input.new_empty(torch.library.get_ctx().new_dynamic_size())

def fn(x):
    return torch.ops.my_custom_library.custom_op(x, torch.tensor([1, 1, 1, 1], dtype=torch.int64))

aot_autograd_check(fn, (torch.randn(10, requires_grad=True),), {}, True)
```

Error:
```bash
Traceback (most recent call last):
  File "/data/users/williamwen/pytorch/playground2.py", line 41, in <module>
    aot_autograd_check(fn, (torch.randn(10, requires_grad=True),), {}, True)
  File "/data/users/williamwen/pytorch/torch/testing/_internal/optests/aot_autograd.py", line 76, in aot_autograd_check
    _test_aot_autograd_forwards_backwards_helper(
  File "/data/users/williamwen/pytorch/torch/testing/_internal/optests/aot_autograd.py", line 137, in _test_aot_autograd_forwards_backwards_helper
    check(args, ignore_failure=False)
  File "/data/users/williamwen/pytorch/torch/testing/_internal/optests/aot_autograd.py", line 133, in check
    compiled_out, compiled_grad = call_forwards_backwards(compiled_f, args)
  File "/data/users/williamwen/pytorch/torch/testing/_internal/optests/aot_autograd.py", line 97, in call_forwards_backwards
    out = wrapper_set_seed(f, args)
  File "/data/users/williamwen/pytorch/torch/testing/_internal/common_methods_invocations.py", line 9245, in wrapper_set_seed
    return op(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 3758, in returned_function
    compiled_fn = create_aot_dispatcher_function(
  File "/data/users/williamwen/pytorch/torch/_dynamo/utils.py", line 221, in time_wrapper
    r = func(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 3462, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 2243, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 2423, in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 2837, in aot_dispatch_autograd
    fx_g = aot_dispatch_autograd_graph(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 2814, in aot_dispatch_autograd_graph
    fx_g = create_functionalized_graph(
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 1451, in create_functionalized_graph
    fx_g = make_fx(helper, decomposition_table=aot_config.decompositions)(*args)
  File "/data/users/williamwen/pytorch/torch/fx/experimental/proxy_tensor.py", line 858, in wrapped
    t = dispatch_trace(wrap_key(func, args, fx_tracer, pre_dispatch), tracer=fx_tracer, concrete_args=tuple(phs))
  File "/data/users/williamwen/pytorch/torch/_compile.py", line 24, in inner
    return torch._dynamo.disable(fn, recursive)(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/_dynamo/eval_frame.py", line 401, in _fn
    return fn(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/fx/experimental/proxy_tensor.py", line 478, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
  File "/data/users/williamwen/pytorch/torch/fx/_symbolic_trace.py", line 817, in trace
    (self.create_arg(fn(*args)),),
  File "/data/users/williamwen/pytorch/torch/fx/_symbolic_trace.py", line 684, in flatten_fn
    tree_out = root_fn(*tree_args)
  File "/data/users/williamwen/pytorch/torch/fx/experimental/proxy_tensor.py", line 514, in wrapped
    out = f(*tensors)
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 1440, in joint_helper
    return functionalized_f_helper(primals, tangents)
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 1396, in functionalized_f_helper
    f_outs = fn(*f_args)
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 1365, in inner_fn_with_anomaly
    return inner_fn(*args)
  File "/data/users/williamwen/pytorch/torch/_functorch/aot_autograd.py", line 1327, in inner_fn
    out if out.shape == tangent.shape else out.view(tangent.shape)
  File "/data/users/williamwen/pytorch/torch/__init__.py", line 361, in __bool__
    return self.node.bool_()
  File "/data/users/williamwen/pytorch/torch/fx/experimental/symbolic_shapes.py", line 1052, in bool_
    return self.guard_bool("", 0)
  File "/data/users/williamwen/pytorch/torch/fx/experimental/symbolic_shapes.py", line 1022, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint, fx_node=self.fx_node)
  File "/data/users/williamwen/pytorch/torch/fx/experimental/recording.py", line 250, in wrapper
    return event.run(self)
  File "/data/users/williamwen/pytorch/torch/fx/experimental/recording.py", line 151, in run
    return self.f(*args, **kwargs)
  File "/data/users/williamwen/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3986, in evaluate_expr
    concrete_val = self.size_hint(orig_expr)
  File "/data/users/williamwen/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3780, in size_hint
    raise self._make_data_dependent_error(result_expr, expr)
torch.fx.experimental.symbolic_shapes.GuardOnDataDependentSymNode: It appears that you're trying to get a value out of symbolic int/float whose value is data-dependent (and thus we do not know the true value.)  The expression we were trying to evaluate is Eq(i1, i0) (unhinted: Eq(i1, i0)).  Scroll up to see where each of these data-dependent accesses originally occurred.
```

If we remove the dynamic shapes, then the script runs fine:
```python
import torch
from torch.library import Library, impl
from torch.testing._internal.optests import aot_autograd_check

m = Library("my_custom_library", "DEF")
m.define("custom_op(Tensor input, Tensor lengths) -> Tensor")

class CustomOp(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, lengths):
        ctx.save_for_backward(input, lengths)
        with torch._C._AutoDispatchBelowAutograd():
            return torch.ops.my_custom_library.custom_op(input, lengths)

    @staticmethod
    def backward(ctx, grad):
        input, lengths = ctx.saved_tensors
        # idx = lengths.sum()
        idx = 4
        grad_out = torch.zeros_like(input)
        grad_out[:idx] = grad
        return grad_out, None

@impl(m, "custom_op", "CPU")
def custom_op_impl(input, lengths):
    # idx = lengths.sum()
    idx = 4
    return input[:idx].clone()

@impl(m, "custom_op", "Autograd")
def custom_op_autograd(input, lengths):
    return CustomOp.apply(input, lengths)

@torch.library.impl_abstract("my_custom_library::custom_op")
def custom_op_meta(input, lengths):
    # return input.new_empty(torch.library.get_ctx().new_dynamic_size())
    return input.new_empty(4)

def fn(x):
    return torch.ops.my_custom_library.custom_op(x, torch.tensor([1, 1, 1, 1], dtype=torch.int64))

aot_autograd_check(fn, (torch.randn(10, requires_grad=True),), {}, True)
```

cc @chauhang @penguinwu @ezyang @bobrenjc93 @aditvenk @laithsakka @bdhirsh @aorenste @gchanan @zou3519 @kadeng @msaroufim @wconstab @anijain2305

**Comment by zou3519** [comment]
TL;DR of what's going on:
- AOTAutograd traces the function one time to get outputs. Then, it constructs grad_output based on these output.
- Then, AOTAutograd traces the function's forward a second time. It checks if the function's forward output has the same shape as the grad_output.
- Unfortunately, we allocate a fresh symint for the operator's output each time we trace it. Comparing the two symints leads to the above error.

thoughts? cc @ezyang

**Comment by jon-chuang** [comment]
@zou3519 is this a higher order ops thing?

**Comment by zou3519** [comment]
No, the case we ran into it was when @williamwen42 wrote the abstract impl for a custom op that returned a dynamic-shape output that requires grad

**Comment by ezyang** [comment]
> Unfortunately, we allocate a fresh symint for the operator's output each time we trace it. Comparing the two symints leads to the above error.

I have noticed this problem and it is a pain. I don't have good ideas for how to fix it, besides maybe aggressive memoization, but this is pretty hard to implement correctly.

**Comment by jon-chuang** [comment]
@ezyang we could create a `symnode_context`? :thinking: 

If the code is related and we need it to share symints, it should be too difficult to pipe around a shared context? 

Though, isn't `ShapeEnv` the right context to pipe around here?

**Comment by ezyang** [comment]
The ShapeEnv is kept around, the problem is that we don't know the new call to nonzero should give the same unbacked symint as the old call

**Comment by jon-chuang** [comment]
Strangely enough, I think the solution may have a lot in common with https://github.com/pytorch/pytorch/issues/110700 - we want to pair the outs and tangents, and propagate some shared information between them (name, shape)

**Comment by ezyang** [comment]
Yes, so retracing under controlled conditions could be made to work, by saying something like "if you call nonzero with these arguments at this particular spot, use this unbacked symint", aka aggressive memoization.

**Comment by bdhirsh** [comment]
Sounds like the root cause of this issue is https://github.com/pytorch/pytorch/issues/110136 ?

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/pull/114368 is an equivalent fix for FakeTensorProp. We want to do something similar here.

**Comment by ezyang** [comment]
So I've landed / have some fixes in progress, but it turns out they don't fix this problem!  Here's the stacktrace of the unbacked SymInt allocation that is duped:

<img width="1118" alt="image" src="https://github.com/pytorch/pytorch/assets/13564/6e3b6984-ade6-45cc-91aa-afae72672646">

Notably, it is happening in *backwards*. The thing I implemented forces sizes to line up in forwards, but not in backwards.

**Comment by ezyang** [comment]
So, it turns out that you can write your backwards function carefully so that you don't have this problem:

```
import torch
from torch.library import Library, impl
from torch.testing._internal.optests import aot_autograd_check

m = Library("my_custom_library", "DEF")
m.define("custom_op(Tensor input, Tensor lengths) -> Tensor")

class CustomOp(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, lengths):
        ctx.save_for_backward(input, lengths)
        with torch._C._AutoDispatchBelowAutograd():
            return torch.ops.my_custom_library.custom_op(input, lengths)

    @staticmethod
    def backward(ctx, grad):
        input, lengths = ctx.saved_tensors
        idx = lengths.sum()
        grad_out = torch.zeros_like(input)
        # NEW STUFF
        i = idx.item()
        torch._check_is_size(i)
        torch._check(i == grad.size(0))
        grad_out.narrow(0, 0, i).copy_(grad)
        # OLD STUFF
        # grad_out[:idx] = grad
        return grad_out, None

@impl(m, "custom_op", "CPU")
def custom_op_impl(input, lengths):
    idx = lengths.sum()
    return input[:idx].clone()

@impl(m, "custom_op", "Autograd")
def custom_op_autograd(input, lengths):
    return CustomOp.apply(input, lengths)

@torch.library.impl_abstract("my_custom_library::custom_op")
def custom_op_meta(input, lengths):
    return input.new_empty(torch.library.get_ctx().new_dynamic_size())

def fn(x):
    return torch.ops.my_custom_library.custom_op(x, torch.tensor([1, 1, 1, 1], dtype=torch.int64))

aot_autograd_check(fn, (torch.randn(10, requires_grad=True),), {}, True)
```

Specifically, I replaced:

```
grad_out[:idx] = grad
```

with

```
        i = idx.item()
        torch._check_is_size(i)
        torch._check(i == grad.size(0))
        grad_out.narrow(0, 0, i).copy_(grad)
```

There are two problems. The first problem is that you can't use slice indexing in code that will be directly make_fx traced, because CPython will unconditionally force specialization here (this is unavoidable, `__index__` must return an int, and `slice()` always immediately invokes this on its inputs.) Using narrow instead is not too difficult.

The next problem is a disagreement about the unbacked SymInts. What happened is that we are actually allocating an unbacked SymInt representing lengths.sum() twice: once in the custom_op_meta, and once implicitly when we called `idx.item()`. In fact, there is *no way* (even with memoization) for us to know a priori that they are equal in this example, because `custom_op_meta` is generating its unbacked SymInt out of thin air so we don't know it's related to lengths in a particular way.

Whenever you have a problem like this, usually it works to just help out with some more asserts. It looks like I had to do two. I think it may not be too difficult to make narrow automatically apply the `check_is_size`; I'm less sure about the equality test.

**Comment by ezyang** [comment]
I'm reducing priority because this is a user error. Maybe we should keep this open for fixes to (1) make it easier to diagnose and (2) sand over some rough edges.

**Comment by ezyang** [comment]
The new error seems better, I guess?

```
  File "/data/users/ezyang/a/pytorch/torch/autograd/function.py", line 315, in apply
    return user_fn(self, *args)
  File "/data/users/ezyang/a/pytorch/all.py", line 20, in backward
    grad_out[:idx] = grad
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/sym_node.py", line 518, in guard_int
    r = self.evaluate()
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/sym_node.py", line 512, in evaluate
    return self.shape_env.evaluate_sym_node(self, size_oblivious)
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/symbolic_shapes.py", line 7239, in evaluate_sym_node
    return self.evaluate_expr(
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/symbolic_shapes.py", line 7339, in evaluate_expr
    return self._inner_evaluate_expr(
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/recording.py", line 272, in wrapper
    return retlog(fn(*args, **kwargs))
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/symbolic_shapes.py", line 7362, in _inner_evaluate_expr
    return self._evaluate_expr(
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/symbolic_shapes.py", line 7586, in _evaluate_expr
    raise self._make_data_dependent_error(
torch.fx.experimental.symbolic_shapes.GuardOnDataDependentSymNode: Could not extract specialized integer from data-dependent expression u2 (unhinted: u2).  (Size-like symbols: none)

Caused by: (all.py:20 in backward)
For more information, run with TORCH_LOGS="dynamic"
For extended logs when we create symbols, also add TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="u2"
If you suspect the guard was triggered from C++, add TORCHDYNAMO_EXTENDED_DEBUG_CPP=1
For more debugging help, see https://docs.google.com/document/d/1HSuTTVvYH1pTew89Rtpeu84Ht3nQEFTYhAX3Ypa_xJs/edit?usp=sharing

```
